core/src/main/java/org/apache/iceberg/AbstractTableScan.java

```diff
@@ -0,0 +1,173 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.events.Listeners;
+import org.apache.iceberg.events.ScanEvent;
+import org.apache.iceberg.expressions.ExpressionUtil;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.metrics.DefaultMetricsContext;
+import org.apache.iceberg.metrics.ImmutableScanReport;
+import org.apache.iceberg.metrics.ScanMetrics;
+import org.apache.iceberg.metrics.ScanMetricsResult;
+import org.apache.iceberg.metrics.ScanReport;
+import org.apache.iceberg.metrics.Timer;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.util.DateTimeUtil;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.iceberg.util.TableScanUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractTableScan<ThisT, T extends ScanTask, C extends ScanTaskGroup<T>>
+    extends BaseScan<ThisT, T, C> {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractTableScan.class);
+  private ScanMetrics scanMetrics;
+
+  protected AbstractTableScan(
+      TableOperations ops, Table table, Schema schema, TableScanContext context) {
+    super(ops, table, schema, context);
+  }
+
+  protected Long snapshotId() {
+    return context().snapshotId();
+  }
+
+  protected Map<String, String> options() {
+    return context().options();
+  }
+
+  protected abstract CloseableIterable<T> doPlanFiles();
+
+  protected ScanMetrics scanMetrics() {
+    if (scanMetrics == null) {
+      this.scanMetrics = ScanMetrics.of(new DefaultMetricsContext());
+    }
+
+    return scanMetrics;
+  }
+
+  @Override
+  public Table table() {
+    return super.table();
+  }
+
+  public ThisT useSnapshot(long scanSnapshotId) {
+    Preconditions.checkArgument(
+        snapshotId() == null, "Cannot override snapshot, already set snapshot id=%s", snapshotId());
+    Preconditions.checkArgument(
+        tableOps().current().snapshot(scanSnapshotId) != null,
+        "Cannot find snapshot with ID %s",
+        scanSnapshotId);
+    return newRefinedScan(
+        tableOps(), table(), tableSchema(), context().useSnapshotId(scanSnapshotId));
```

Comment:
nit: One of the things we did in `BaseIncrementalScan` is to create a temp var for the new context to avoid splitting this call into multiple lines.

```
TableScanContext newContext = context().useSnapshotId(scanSnapshotId);
return newRefinedScan(tableOps(), table(), tableSchema(), newContext);
```

Reply:
Done